### DISCUSSION TITLE
How to send error to client in streaming responses?

### DISCUSSION THREAD

**Discussion by liss-h** [tokio-rs/axum#3042] [Q&A] [answered]
### Summary

Hello,
I was wondering if there is any way to return an error to the client in a streaming response **after** the stream has started.

Imagine something like this:
```rust
async fn handle_request(...) -> impl IntoResponse {
    let stream: impl Stream<Item = io::Result<Bytes>> = /* some stream that produces an error */;
    Body::from_stream(stream)
}
```

Now, this does not exactly do what I expected: In the case the stream produces an `Err(_)` axum aborts the connection.

Is there any way to send the error to the client instead?
With my very surface-level HTTP knowledge it would seem that this is not possible with HTTP chunked transfer encoding in general, but maybe you have some suggestion to address this some other way.

Note: Our usecase is a triplestore with `SPARQL`. We do not know in advance how much data we'll produce and we also can't know if or when an error will occur during query evaluation.

Thanks in advance 🙂

### axum version

0.7.7 atm, but we can upgrade to a newer one if required

**Comment by mladedav** [comment, marked as answer]
Sorry, that's not possible. You have to first send a status code (in your case most likely 200 since you're not expecting errors yet) and then you start streaming the body. At that time you can't change your mind to sending 400 instead.

Closing the connection is pretty much the only standard way to signal to the client that something has gone wrong.

If you're controlling both the client and server, you can also utilize trailers after the body to signal errors or any other metadata.

#1951 might be of some interest to you.
